### DISCUSSION TITLE
How come this package is not stable?

### DISCUSSION THREAD

**Discussion by lindeberg** [appany/AppAny.HotChocolate.FluentValidation#70] [Q&A] [answered]
@sergeyshaykhullin you [mentioned](https://github.com/appany/AppAny.HotChocolate.FluentValidation/issues/67#issuecomment-831663936) that this package is not stable. What do you mean by that? Is it production ready? If not, what is missing or what are the issues?

**Comment by sergeyshaykhullin** [comment, marked as answer]
You can use it is production, i cover a lot of use cases and have 99%+ test coverage. I am using it every day and for now have no troubles and new bugs

This is package is not stable, because i am not sure in public api stability:
- I have one issue which i can't close without a lot of internal rewrites and may be public api breaking changes
- I have some closed prs with ideas to migrate to strongly typed errors
- I want to migrate to middleware approach to compose and reuse code
- Also i am using some parts of functional programming inside and have some space for improvements and reimplementations

This is my internal roadmap and mostly "nice to haves"

I mentioned FairyBread, it has already 6 major version, but package was released less than year ago

I prefer Rust approach, when 1.0.0 is stable version with stable apis. A lot of pretty decent Rust crates has 0.x versions - no commitment to not to have breaking changes in public apis

**Reply by lindeberg** [reply]
Okay. Makes sense! Interesting to hear your internal roadmap.

I'm myself confused by the many ways and places and types of errors in GraphQL.

Are you considering not using the "errors.extensions" property and placing these errors in the payload instead? Makes sense in some way since they are user errors I guess.

**Reply by sergeyshaykhullin** [reply]
Yes, i have this... Dirty wip prs. It is possible, but should be redesigned
https://github.com/appany/AppAny.HotChocolate.FluentValidation/pull/28
https://github.com/appany/AppAny.HotChocolate.FluentValidation/pull/29

**Reply by lindeberg** [reply]
Interesting! There is no standard for this?

**Reply by sergeyshaykhullin** [reply]
No

**Reply by lindeberg** [reply]
But you now have an opinion towards putting it on the payload? Have you disliked putting this in the GQL-errors field?
What about Authentication? :D

**Reply by sergeyshaykhullin** [reply]
It is now in errors field. The main problem is that errors has no well-defined schema = not strongly typed and something like apollo-client or StrawberryShake can't generate models for them

**Reply by lindeberg** [reply]
Found this in the HC-slack
![image](https://user-images.githubusercontent.com/15940340/116995933-fea4e700-acda-11eb-81df-1a7315e8413b.png)

**Reply by sergeyshaykhullin** [reply]
It is not a problem, all HC validation libraries is using errors, but i did some r&d and discovered alternative approach based on HotChocolate

**Reply by lindeberg** [reply]
Why need to have types? What would the types be? "FirstNameNullOrWhitespaceError"? A ValidationError with 
Code
Message
Field
is fine?

**Reply by sergeyshaykhullin** [reply]
You can see draft prs for more details. You need types for results to generate... Ok(payload) or Error(validation details) with the same shape as input parameters
